File: tests/test_field_control.py
"""Authored source-branch fixtures, separate from local original-run evidence."""

import unittest
from dataclasses import replace

from tools.analysis.events import (
    EventError,
    UnknownInstruction,
    decode_instruction,
    disassemble_reachable,
)
from tools.analysis.field_control import (
    ControlActor,
    ControlError,
    ControlInputs,
    ControlState,
    EncounterGates,
    UnrecoveredEncounter,
    encounter_early_exit,
    request_control,
    terrain_jump_result,
)


def actor(**changes):
    return replace(ControlActor(0x4000, 0, (12, -5, 31), (12, -5, 31), 0x8123, 7, 123), **changes)


def state(**changes):
    return replace(ControlState(0, 0, 987, 0, 9), **changes)


def inputs(**changes):
    # Invented direction tables; original tables are imported only for local comparison.
    tables = (tuple(range(0, 4096, 256)), tuple(0x8000 + i for i in range(16)))
    gates = EncounterGates(1, 1, 1, 0, 0, 0, 0, 1)
    return replace(
        ControlInputs(0, 0, (-1, -1, -1, -1), 0, 255, 0, 8, 42, 0, tables, 0, gates), **changes
    )


class FieldControlTests(unittest.TestCase):
    def test_non_owner_preserves_direction_and_control_state(self):
        before = actor(flags=0xA0040080, pc=65535)
        result = request_control(0xA7, before, state(), inputs(pressed_buttons=0x80))
        self.assertEqual(result.actor, replace(before, flags=0xA1040080, pc=0))
        self.assertEqual(result.state, state())
        self.assertEqual((result.eligibility, result.calls), ("not_control_owner", ()))
        result = request_control(0xA7, before, state(), inputs(preserve_nonplayer_motion=2))
        self.assertEqual(result.actor, replace(before, pc=0))

    def test_dialogue_and_inhibition_block_before_encounter_and_jump(self):
        for index in range(4):
            statuses = [-1] * 4
            statuses[index] = 0
            result = request_control(
                0xA7,
                actor(),
                state(),
                inputs(dialogue_status=tuple(statuses), held_buttons=0x1080, pressed_buttons=0x80),
            )
            self.assertEqual(result.actor, replace(actor(), direction=0x8000, pc=124))
            self.assertEqual(result.state, state())
            self.assertEqual((result.eligibility, result.calls), ("dialogue", ()))
        for inhibition in (-32768, -1, 1, 32767):
            result = request_control(
                0xA7,
                actor(),
                state(),
                inputs(encounter=replace(inputs().encounter, inhibition=inhibition)),
            )
            self.assertEqual((result.eligibility, result.actor.direction), ("inhibited", 0x8000))

    def test_pressed_and_held_jump_are_distinct(self):
        result = request_control(0xA7, actor(), state(), inputs(pressed_buttons=0x80))
        self.assertEqual(
            (result.actor.flags, result.jump, result.calls), (0x4800, "pressed", ("terrain",))
        )
        self.assertEqual((result.state.latched_jump_setting, result.state.updated), (42, 1))
        held = request_control(0xA7, actor(), state(), inputs(held_buttons=0x80))
        self.assertEqual(
            (held.actor.flags, held.jump, held.state.latched_jump_setting),
            (0x4000, "not_pressed", 987),
        )

    def test_normal_jump_rejections_have_no_jump_side_effect(self):
        cases = [
            (actor(flags=0x4800), inputs(pressed_buttons=0x80), "jump_or_airborne"),
            (actor(flags=0x5000), inputs(pressed_buttons=0x80), "jump_or_airborne"),
            (actor(terrain_flags=0x400000), inputs(pressed_buttons=0x80), "counter_terrain"),
            (actor(), inputs(pressed_buttons=0x80, jump_contact=0x100FF), "contact"),
            (actor(flags=0x4200, terrain_flags=8), inputs(pressed_buttons=0x80), "terrain_mask"),
        ]
        for before, controls, reason in cases:
            result = request_control(0xA7, before, state(), controls)
            self.assertEqual(
                (result.actor.flags, result.state.latched_jump_setting), (before.flags, 987)
            )
            self.assertEqual(result.jump, reason)

    def test_terrain_predicate_matches_only_two_corresponding_bits(self):
        for flags, terrain, result in (
            (0, 24, 0),
            (0x200, 8, -1),
            (0x200, 16, 0),
            (0x400, 8, 0),
            (0x400, 16, -1),
            (0x600, 24, -1),
            (0x800, 32, 0),
            (0xFFFFFFFF, 0xFFFFFFE7, 0),
        ):
            self.assertEqual(terrain_jump_result(actor(flags=flags, terrain_flags=terrain)), result)

    def test_counter_held_retry_threshold_and_retention_on_movement(self):
        before = actor(terrain_flags=0x400000)
        controls = inputs(held_buttons=0x80)
        first = request_control(0xA7, before, state(stationary_counter=31), controls)
        self.assertEqual((first.state.stationary_counter, first.jump), (32, "not_pressed"))
        second = request_control(0xA7, before, first.state, controls)
        self.assertEqual((second.state.stationary_counter, second.jump), (32, "held_retry"))
        moving = replace(before, integer_position=(13, -5, 31))
        result = request_control(0xA7, moving, state(stationary_counter=17), controls)
        self.assertEqual(result.state.stationary_counter, 17)
        result = request_control(0xA7, moving, state(stationary_counter=33), controls)
        self.assertEqual((result.state.stationary_counter, result.jump), (32, "held_retry"))
        self.assertEqual(
            request_control(0xA7, actor(), second.state, controls).state.stationary_counter, 0
        )

    def test_counter_wraps_before_signed_threshold_comparison(self):
        for value, expected in ((0x7FFF, 0x8000), (0xFFFE, 0xFFFF), (0xFFFF, 0)):
            result = request_control(
                0xA7,
                actor(terrain_flags=0x400000),
                state(stationary_counter=value),
                inputs(held_buttons=0x80),
            )
            self.assertEqual(
                (result.state.stationary_counter, result.jump), (expected, "not_pressed")
            )

    def test_alternate_mode_reloads_then_decrements_without_airborne_gate(self):
        controls = inputs(jump_mode=-1, pressed_buttons=0x80, repeat_delay=2)
        result = request_control(0xA7, actor(flags=0x5800), state(), controls)
        self.assertEqual(
            (result.jump, result.actor.animation_mode, result.state.repeat_remaining),
            ("pressed", 255, 1),
        )
        blocked = request_control(0xA7, actor(), result.state, controls)
        self.assertEqual(
            (blocked.jump, blocked.state.repeat_remaining, blocked.calls), ("repeat_delay", 0, ())
        )
        empty = request_control(0xA7, actor(), state(), replace(controls, repeat_delay=0))
        self.assertEqual(empty.state.repeat_remaining, 0)
        negative = request_control(
            0xA7, actor(), state(repeat_remaining=65535), inputs(jump_mode=1)
        )
        self.assertEqual(negative.state.repeat_remaining, 65534)

    def test_held_retry_bypasses_alternate_mode_even_when_terrain_rejects(self):
        for terrain, jump in ((0x400000, "held_retry"), (0x400008, "terrain_mask")):
            result = request_control(
                0xA7,
                actor(flags=0x4200, terrain_flags=terrain),
                state(stationary_counter=32, repeat_remaining=7),
                inputs(jump_mode=1, held_buttons=0x80),
            )
            self.assertEqual((result.jump, result.terrain_call_variant), (jump, "normal"))
            self.assertEqual((result.actor.animation_mode, result.state.repeat_remaining), (7, 7))

    def test_direction_uses_all_nibble_combinations_and_preserves_sentinel_bits(self):
        expected = (
            0xEEE,
            0xDEE,
            0xCEE,
            0xBEE,
            0xAEE,
            0x9EE,
            0x8EE,
            0x7EE,
            0x6EE,
            0x5EE,
            0x4EE,
            0x3EE,
            0x2EE,
            0x1EE,
            0x0EE,
            0xFEE,
        )
        for nibble, direction in enumerate(expected):
            result = request_control(
                0xA7, actor(), state(), inputs(held_buttons=nibble << 12, camera_angle=18)
            )
            self.assertEqual(result.actor.direction, direction)
            sentinel = request_control(
                0xA7,
                actor(),
                state(),
                inputs(held_buttons=nibble << 12, camera_angle=-32768, alternate_directions=2),
            )
            self.assertEqual(sentinel.actor.direction, 0x800F - nibble)

    def test_encounter_gates_use_exact_values_and_first_return(self):
        open_gates = replace(inputs().encounter, encounter_gate=1, menu_gate=2)
        self.assertIsNone(encounter_early_exit(open_gates))
        for name, value, reason in (
            ("field_active", 0, "field_inactive"),
            ("gate_e4", 0, "gate_e4_clear"),
            ("gate_ec", 0, "gate_ec_clear"),
            ("music_result", 0xFFFFFFFF, "music_pending"),
            ("encounter_gate", 0, "encounter_gate_clear"),
            ("inhibition", -1, "inhibited"),
            ("menu_gate", 1, "menu_gate_one"),
            ("enabled_byte", 0, "enabled_byte_clear"),
        ):
            self.assertEqual(encounter_early_exit(replace(open_gates, **{name: value})), reason)
        self.assertIsNone(
            encounter_early_exit(replace(open_gates, inhibition=1, music_result=0xFFFFFFFE))
        )
        self.assertEqual(
            encounter_early_exit(replace(open_gates, field_active=0, gate_e4=0)), "field_inactive"
        )
        with self.assertRaises(UnrecoveredEncounter):
            request_control(
                0xA7, actor(), state(), inputs(held_buttons=0x1000, encounter=open_gates)
            )
        self.assertEqual(
            request_control(0xA7, actor(), state(), inputs(encounter=open_gates)).calls, ()
        )

    def test_looping_wrapper_restores_pc_and_requests_break_even_when_blocked(self):
        for before in (actor(pc=65535), actor(flags=0, pc=65535)):
            result = request_control(0x0C, before, state(), inputs(pressed_buttons=0x80))
            plain = request_control(0xA7, before, state(), inputs(pressed_buttons=0x80))
            self.assertEqual(result.actor, replace(plain.actor, pc=65535))
            self.assertEqual(result.state, replace(plain.state, break_requested=1))

    def test_invalid_widths_and_unrecovered_opcode_fail_explicitly(self):
        for invalid in (-1, 65536, True):
            with self.assertRaises(ControlError):
                inputs(held_buttons=invalid)
        for invalid in (-32769, 32768):
            with self.assertRaises(ControlError):
                actor(integer_position=(invalid, 0, 0))
        with self.assertRaises(ControlError):
            inputs(direction_tables=((0,) * 15, (0,) * 16))
        with self.assertRaises(ControlError):
            request_control(0x0B, actor(), state(), inputs())

    def test_control_instruction_lengths_and_reachable_successors(self):
        plain = decode_instruction(bytes([0xA7, 0]), 0)
        self.assertEqual((plain.size, plain.operands, plain.successors), (1, (), (1,)))
        loop = decode_instruction(bytes([0x0C]), 0)
        self.assertEqual((loop.size, loop.operands, loop.successors), (1, (), (0,)))
        self.assertEqual(disassemble_reachable(bytes([0x0C]), 0), (loop,))
        self.assertEqual([i.pc for i in disassemble_reachable(bytes([0xA7, 0]), 0)], [0, 1])
        wrapped = decode_instruction(bytes(65535) + bytes([0xA7]), 65535)
        self.assertEqual(wrapped.successors, (0,))
        with self.assertRaises(EventError):
            decode_instruction(bytes([0xA7]), 0)
        with self.assertRaises(UnknownInstruction):
            decode_instruction(bytes([0x0B, 0]), 0)


if __name__ == "__main__":
    unittest.main()

File: tests/test_field_control.cpp
#include "xem/reconstruction/field_control.hpp"

#include <array>
#include <iostream>
#include <vector>

namespace field = xem::reconstruction::field;
namespace {
void check(bool value, const char *message) {
    if (!value)
        throw std::runtime_error(message);
}
template <typename Error, typename Operation>
void rejects(Operation operation, const char *message) {
    try {
        operation();
    } catch (const Error &) {
        return;
    }
    throw std::runtime_error(message);
}

// Invented tables and inputs. Original control traces are compared separately.
struct Scenario {
    field::EventActor event_actor;
    field::ControlActor actor{0, {12, -5, 31}, {12, -5, 31}, 0x8123, 7};
    field::ControlState state{0, 0, 987, 0};
    field::ControlInputs inputs;
    field::BattleRequestState battle{1, 0, 0, 0, 0, 0, 0};
    field::EventContext context;
    Scenario() {
        event_actor.flags = 0x4000;
        event_actor.pc = 123;
        event_actor.selected_slot = 255; // These source handlers never read a selected slot.
        context.current_actor = &event_actor;
        context.control.gate_values = {1, 1, 1};
        context.control.break_requested = 9;
        inputs.dialogue_status = {-1, -1, -1, -1};
        inputs.jump_contact = 255;
        inputs.repeat_delay = 8;
        inputs.jump_setting = 42;
        inputs.encounter.enabled_byte = 1;
        for (std::uint16_t i = 0; i < 16; ++i) {
            inputs.direction_tables[0][i] = static_cast<std::uint16_t>(i * 256U);
            inputs.direction_tables[1][i] = static_cast<std::uint16_t>(0x8000U + i);
        }
    }
    field::ControlEffect run() {
        return field::request_field_control(context, actor, state, inputs, battle);
    }
};

void ownership_and_gates() {
    for (const auto preserve : {std::uint8_t{0}, std::uint8_t{2}}) {
        Scenario s;
        s.event_actor.flags = 0xa0040080;
        s.event_actor.pc = 65535;
        s.inputs.preserve_nonplayer_motion = preserve;
        s.inputs.held_buttons = 0x1080;
        s.inputs.pressed_buttons = 0x80;
        s.inputs.encounter.encounter_gate = 1; // Would be unsupported if reached.
        const auto before = s.actor;
        const auto state = s.state;
        const auto effect = s.run();
        check(s.event_actor.flags == (preserve == 0 ? 0xa1040080U : 0xa0040080U) &&
                  s.event_actor.pc == 0 && s.actor == before && s.state == state &&
                  effect.eligibility == field::ControlEligibility::not_control_owner &&
                  effect.call_count == 0 && s.context.control.break_requested == 9,
              "Non-owner path preserves control data and wraps PC after optional motion flag");
    }
    for (std::size_t blocked = 0; blocked < 4; ++blocked) {
        Scenario s;
        s.inputs.dialogue_status[blocked] = 0;
        s.inputs.held_buttons = 0x1080;
        s.inputs.encounter.encounter_gate = 1;
        const auto before = s.state;
        const auto effect = s.run();
        check(effect.eligibility == field::ControlEligibility::dialogue && effect.call_count == 0 &&
                  s.actor.direction == 0x8000 && s.state == before && s.event_actor.pc == 124,
              "Each of four zero dialogue statuses blocks before nested calls");
    }
    for (const auto inhibition : std::array<std::int16_t, 4>{-32768, -1, 1, 32767}) {
        Scenario s;
        s.inputs.encounter.inhibition = inhibition;
        check(s.run().eligibility == field::ControlEligibility::inhibited &&
                  s.actor.direction == 0x8000,
              "Any nonzero control inhibition blocks ownership");
    }
    for (unsigned first = 0; first < 8; ++first) {
        Scenario s;
        const std::array reasons{
            field::EncounterExit::field_inactive,       field::EncounterExit::gate_e4_clear,
            field::EncounterExit::gate_ec_clear,        field::EncounterExit::music_pending,
            field::EncounterExit::encounter_gate_clear, field::EncounterExit::inhibited,
            field::EncounterExit::menu_gate_one,        field::EncounterExit::enabled_byte_clear};
        s.inputs.encounter.encounter_gate = 1;
        if (first == 0)
            s.battle.field_active = 0;
        if (first <= 1)
            s.context.control.gate_values[1] = 0;
        if (first <= 2)
            s.context.control.gate_values[2] = 0;
        if (first <= 3)
            s.inputs.encounter.music_result = 0xffffffffU;
        if (first <= 4)
            s.inputs.encounter.encounter_gate = 0;
        if (first <= 5)
            s.inputs.encounter.inhibition = -1;
        if (first <= 6)
            s.battle.menu_gate = 1;
        s.inputs.encounter.enabled_byte = 0;
        check(field::encounter_early_exit(s.battle, s.context.control, s.inputs.encounter) ==
                  reasons[first],
              "Encounter returns must test the original ordered gates");
    }
    Scenario active;
    active.inputs.encounter.encounter_gate = 1;
    active.inputs.encounter.music_result = 0xfffffffeU;
    active.battle.menu_gate = 2;
    check(field::encounter_early_exit(active.battle, active.context.control,
                                      active.inputs.encounter) == field::EncounterExit::none,
          "Only exact menu-one and music-minus-one values block the encounter poll");
    active.inputs.held_buttons = 0x1000;
    const auto before = active.actor;
    const auto state = active.state;
    rejects<field::UnrecoveredEncounter>([&] { (void)active.run(); },
                                         "Active encounter selection must fail explicitly");
    check(active.actor == before && active.state == state && active.event_actor.pc == 123 &&
              active.context.control.break_requested == 9,
          "Unrecovered encounter must not publish downstream control changes");
    active.inputs.held_buttons = 0;
    check(active.run().call_count == 0, "No direction request means no encounter poll");
}

void normal_jump_and_terrain() {
    Scenario s;
    s.inputs.pressed_buttons = 0x80;
    s.inputs.held_buttons = 0x1000;
    const auto effect = s.run();
    check(s.event_actor.flags == 0x4800 && s.state.latched_jump_setting == 42 &&
              s.state.updated == 1 && effect.jump == field::JumpRequest::pressed &&
              effect.call_count == 2 && effect.calls[0] == field::ControlCall::encounter &&
              effect.calls[1] == field::ControlCall::terrain &&
              effect.terrain_call == field::TerrainCall::normal,
          "Pressed jump follows the real encounter early return and terrain predicate");
    Scenario held;
    held.inputs.held_buttons = 0x80;
    check(held.run().jump == field::JumpRequest::not_pressed && held.event_actor.flags == 0x4000 &&
              held.state.latched_jump_setting == 987,
          "Held Triangle is not a newly pressed normal jump");
    for (unsigned test = 0; test < 5; ++test) {
        Scenario blocked;
        blocked.inputs.pressed_buttons = 0x80;
        const std::array reasons{field::JumpRequest::jump_or_airborne,
                                 field::JumpRequest::jump_or_airborne,
                                 field::JumpRequest::counter_terrain, field::JumpRequest::contact,
                                 field::JumpRequest::terrain_mask};
        if (test < 2)
            blocked.event_actor.flags |= test == 0 ? 0x800U : 0x1000U;
        if (test == 2)
            blocked.actor.terrain_flags = 0x400000;
        if (test == 3)
            blocked.inputs.jump_contact = 0x100ff;
        if (test == 4) {
            blocked.event_actor.flags |= 0x200;
            blocked.actor.terrain_flags = 8;
        }
        const auto flags = blocked.event_actor.flags;
        check(blocked.run().jump == reasons[test] && blocked.event_actor.flags == flags &&
                  blocked.state.latched_jump_setting == 987,
              "Normal jump blockers must preserve flags and latched settings");
    }
    for (std::uint32_t actor_bits = 0; actor_bits < 4; ++actor_bits)
        for (std::uint32_t terrain_bits = 0; terrain_bits < 4; ++terrain_bits)
            check(field::terrain_jump_result(0xfffff9ffU | (actor_bits << 9U),
                                             0xffffffe7U | (terrain_bits << 3U)) ==
                      ((actor_bits & terrain_bits) != 0 ? -1 : 0),
                  "Terrain result must intersect exactly the two corresponding bits");
}

void counter_and_alternate_jump() {
    Scenario s;
    s.actor.terrain_flags = 0x400000;
    s.state.stationary_counter = 31;
    s.inputs.held_buttons = 0x80;
    check(s.run().jump == field::JumpRequest::not_pressed && s.state.stationary_counter == 32,
          "Stationary count equal to 32 must not enable held retry");
    check(s.run().jump == field::JumpRequest::held_retry && s.state.stationary_counter == 32,
          "Stationary count above 32 clamps and enables held retry");
    for (const auto counter :
         {std::uint16_t{0x7fff}, std::uint16_t{0xfffe}, std::uint16_t{0xffff}}) {
        Scenario wrapped;
        wrapped.actor.terrain_flags = 0x400000;
        wrapped.state.stationary_counter = counter;
        wrapped.inputs.held_buttons = 0x80;
        check(wrapped.run().jump == field::JumpRequest::not_pressed &&
                  wrapped.state.stationary_counter == static_cast<std::uint16_t>(counter + 1U),
              "Counter halfword wrap precedes signed threshold comparison");
    }
    Scenario moved;
    moved.actor.terrain_flags = 0x400000;
    moved.actor.integer_position[0] = 13;
    moved.state.stationary_counter = 17;
    (void)moved.run();
    check(moved.state.stationary_counter == 17, "Movement on counter terrain retains its counter");
    moved.actor.terrain_flags = 0;
    (void)moved.run();
    check(moved.state.stationary_counter == 0, "Other terrain resets the stationary counter");

    Scenario alternate;
    alternate.inputs.jump_mode = -1;
    alternate.inputs.pressed_buttons = 0x80;
    alternate.inputs.repeat_delay = 2;
    alternate.event_actor.flags = 0x5800;
    auto effect = alternate.run();
    check(effect.jump == field::JumpRequest::pressed && alternate.actor.animation_mode == 255 &&
              alternate.state.repeat_remaining == 1 &&
              effect.terrain_call == field::TerrainCall::alternate,
          "Alternate jump skips airborne gates, reloads cooldown and decrements immediately");
    effect = alternate.run();
    check(effect.jump == field::JumpRequest::repeat_delay && effect.call_count == 0 &&
              alternate.state.repeat_remaining == 0,
          "Existing alternate cooldown blocks this press before decrementing");
    alternate.inputs.repeat_delay = 0;
    (void)alternate.run();
    check(alternate.state.repeat_remaining == 0, "Reloading zero must not decrement to ffff");
    for (const std::uint32_t terrain : {0x400000U, 0x400008U}) {
        Scenario retry;
        retry.actor.terrain_flags = terrain;
        retry.event_actor.flags = 0x4200;
        retry.state = {32, 7, 987, 0};
        retry.inputs.jump_mode = 1;
        retry.inputs.held_buttons = 0x80;
        effect = retry.run();
        check(effect.terrain_call == field::TerrainCall::normal &&
                  effect.jump == (terrain == 0x400000 ? field::JumpRequest::held_retry
                                                      : field::JumpRequest::terrain_mask) &&
                  retry.state.repeat_remaining == 7 && retry.actor.animation_mode == 7,
              "Held retry bypasses alternate cooldown even when terrain rejects the jump");
    }
}

void directions_and_connected_events() {
    for (unsigned nibble = 0; nibble < 16; ++nibble) {
        Scenario s;
        s.inputs.held_buttons = static_cast<std::uint16_t>(nibble << 12U);
        s.inputs.camera_angle = 18;
        (void)s.run();
        check(s.actor.direction == (((15U - nibble) * 256U - 18U) & 0xfffU),
              "All direction nibbles use XOR 15 and signed camera subtraction");
        s.inputs.camera_angle = -32768;
        s.inputs.alternate_directions = 2;
        (void)s.run();
        check(s.actor.direction == 0x800fU - nibble,
              "Direction sentinel retains every bit and skips camera rotation");
    }
    Scenario s;
    s.event_actor.pc = 0;
    s.event_actor.selected_slot = 0;
    s.context.control.budget_mode = 1;
    field::EventVariables variables;
    s.context.variables = &variables;
    const std::array<std::uint8_t, 8> code{0xa7, 0x35, 0, 0, 9, 0, 0x40, 0};
    s.context.program = {code, {}};
    const auto dispatch = [&](field::EventContext &context, std::uint8_t opcode) {
        if (opcode == 0xa7 || opcode == 0x0c)
            (void)field::execute_control_event(context, opcode, s.actor, s.state, s.inputs,
                                               s.battle);
        else
            field::execute_core_event(context, opcode);
    };
    const auto batch = field::run_event_batch(s.context, 8, dispatch);
    check(batch.dispatched == 3 && batch.reason == field::BatchExit::handler_break &&
              variables.read(0) == 9 && s.event_actor.pc == 7 && s.state.updated == 1,
          "A7 must compose with assignment/end in the real shared event batch");
    std::vector<std::uint8_t> wrapper_code(65536);
    wrapper_code[65535] = 0x0c;
    s.context.program = {wrapper_code, {}};
    s.event_actor.pc = 65535;
    s.event_actor.flags = 0; // Blocked wrapper must still yield without advancing.
    const auto wrapper = field::run_event_batch(s.context, 8, dispatch);
    check(wrapper.dispatched == 1 && wrapper.reason == field::BatchExit::handler_break &&
              s.event_actor.pc == 65535 && s.context.control.break_requested == 1,
          "0C restores the saved PC and requests a break even when blocked");
    wrapper_code[65535] = 0x0b;
    rejects<field::UnsupportedInstruction>(
        [&] {
            (void)field::execute_control_event(s.context, 0x0b, s.actor, s.state, s.inputs,
                                               s.battle);
        },
        "Unknown control opcode must not be treated as a request");
    rejects<field::EventError>(
        [&] {
            (void)field::execute_control_event(s.context, 0xa7, s.actor, s.state, s.inputs,
                                               s.battle);
        },
        "Dispatched control opcode must agree with bytecode");
    s.context.current_actor = nullptr;
    rejects<field::EventError>([&] { (void)s.run(); }, "Missing actor must fail explicitly");
}

void original_actor_import() {
    std::array<std::uint8_t, 0x138> bytes{};
    const auto word = [&](std::size_t offset, std::uint16_t value) {
        bytes[offset] = static_cast<std::uint8_t>(value);
        bytes[offset + 1] = static_cast<std::uint8_t>(value >> 8U);
    };
    word(0x14, 0xff08);
    word(0x16, 0x8040);
    word(0x20, 0x1234); // Fractional position halfwords must not become integer coordinates.
    word(0x22, 0x8000);
    word(0x24, 0x5678);
    word(0x26, 0x7fff);
    word(0x28, 0x9abc);
    word(0x2a, 0xffff);
    word(0x68, 0x7fff);
    word(0x6a, 0x8000);
    word(0x6c, 0xfffe);
    word(0x104, 0x8123);
    word(0xe8, 0xff56);
    const auto state = field::original::read_control_actor(bytes);
    check(state.terrain_flags == 0x8040ff08 &&
              state.integer_position == std::array<std::int16_t, 3>{-32768, 32767, -1} &&
              state.cached_position == std::array<std::int16_t, 3>{32767, -32768, -2} &&
              state.direction == 0x8123 && state.animation_mode == 0xff56,
          "Original actor import must read integer high halves and signed cached coordinates");
}
} // namespace

int main() {
    try {
        ownership_and_gates();
        normal_jump_and_terrain();
        counter_and_alternate_jump();
        directions_and_connected_events();
        original_actor_import();
        std::cout << "Field control: source boundaries and connected event tests passed\n";
        return 0;
    } catch (const std::exception &error) {
        std::cerr << error.what() << '\n';
        return 1;
    }
}
